File: 02_analysis/scripts/05_score_signatures_viz.py
#!/usr/bin/env python
"""
05_score_signatures_viz.py — VIZ (no statistics). The panel that answers the niche question.
===========================================================================================
Reads the fgsea tables + score tables from 05_score_signatures.py and renders:
  - the ordered WT_heat NES dot plot across Treg / Tcon / CD8 — the confirmatory
    answer to whether the mouse 39 °C-derived signature separates the inflamed
    synovial niche from paired blood within a frozen sort label;
  - per-cell WT_heat_up AUCell score violins SF-Treg vs PB-Treg vs SF-Tcon vs SF-CD8
    (running-sum figure moved to 05_score_signatures_viz.R).

Two naming rules bind this file. First, this panel is an ordered NES dot plot:
no `pseudobulk_nes` row carries a standard error or interval. The legacy file
stem `wt_heat_nes_forest` is load-bearing for consumers outside this compartment
and remains unchanged, while reader-facing text names the displayed geometry.
Second, the set is named by how it was derived — from mouse iTreg 37/39 °C
contrasts — and never by a specificity it does not have.
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT))
sys.path.insert(0, str(ROOT / "02_analysis"))
os.chdir(ROOT)

from config import PATHS, PARAMS, POPULATION_COLORS  # noqa: E402
from helpers.figure_style import set_paper_style, save_overview, FIG_CFG  # noqa: E402
from helpers.source_hash_manifest import verify_source_hash  # noqa: E402

STAGE = "05_scoring"
SCRIPT = "02_analysis/scripts/05_score_signatures_viz.py"
POP_TAG = {"Treg": "treg", "Tcon": "tcon", "CD8": "cd8"}
# The one population palette, read from analysis_config.yaml::colors.populations.
POP_COL = POPULATION_COLORS
PRIMARY = "WT_heat"
ARM_ORDER = ["up", "down"]


def nominal_set_sizes() -> dict:
    """Nominal size of each frozen mouse arm — a line count, not a statistic.

    The effective set size only means something against the nominal one, so both
    travel with the panel. Read from the frozen mouse->human projection contract
    that stage 05 scored, so the denominator cannot drift from the numerator.
    """
    sig_dir = PATHS.signature_contract / "signatures" / PRIMARY
    out = {}
    for arm in ARM_ORDER:
        path = sig_dir / f"{PRIMARY}_{arm}.txt"
        if not path.exists():
            raise FileNotFoundError(f"[05_viz] frozen signature not found: {path}")
        verify_source_hash(
            path,
            f"{PRIMARY}_{arm}",
            PATHS.tables(STAGE) / "source_hash_manifest.csv",
            root=ROOT.parent,
        )
        out[arm] = len({ln.strip() for ln in path.read_text().splitlines() if ln.strip()})
    return out


def main() -> None:
    set_paper_style(config=FIG_CFG)
    tdir = PATHS.tables(STAGE)

    # ---- gather NES ----
    rows = []
    for pop, tag in POP_TAG.items():
        f = tdir / f"gsea_pseudobulk_{tag}.csv"
        if f.exists():
            g = pd.read_csv(f)
            g["cell_state"] = pop
            rows.append(g)
    gsea = pd.concat(rows, ignore_index=True) if rows else pd.DataFrame()

    # ---- 1. ordered NES dot plot — the confirmatory answer ----
    fdr = float(PARAMS.gsea_fdr)
    nominal = nominal_set_sizes()
    fig, ax = plt.subplots(figsize=(9.5, 7))
    answer = ""
    if not gsea.empty:
        gsea["n_nominal"] = gsea["direction"].map(nominal)
        order = [(p, d) for d in ARM_ORDER for p in POP_TAG]
        ylabels, yv = [], []
        for i, (pop, d) in enumerate(order):
            sub = gsea[(gsea["cell_state"] == pop) & (gsea["direction"] == d)]
            if sub.empty:
                continue
            nes = float(sub["nes"].iloc[0]); padj = float(sub["padj"].iloc[0])
            n_eff = int(sub["set_size"].iloc[0]); n_nom = int(sub["n_nominal"].iloc[0])
            ax.scatter(nes, i, s=140, color=POP_COL[pop],
                       edgecolors="black", zorder=3,
                       marker="o" if d == "up" else "D")
            star = "*" if padj < fdr else ""
            ax.text(nes, i + 0.20, f"{nes:.2f}{star}", ha="center", fontsize=9)
            # Effective set size travels with every NES, ON THE FACE, against the
            # nominal size of that arm — the project rule, and the guard against
            # reading the ordering of these three NES as a biological ranking.
            ax.text(nes + 0.10, i, f"n {n_eff} of {n_nom}  ·  FDR "
                                   f"{padj:.3f}" if padj >= 0.001 else
                                   f"n {n_eff} of {n_nom}  ·  FDR {padj:.0e}",
                    va="center", ha="left", fontsize=9)
            ylabels.append(f"{pop} · {d}"); yv.append(i)

        # The answer, read off the plotted rows rather than asserted. It is
        # negative on Treg preference, and the panel has to say so on its face.
        up = gsea[gsea["direction"].eq("up")]
        n_sig_up = int(up["padj"].lt(fdr).sum())
        down_sig = gsea[gsea["direction"].eq("down") & gsea["padj"].lt(fdr)]
        answer = (f"pan-T, not Treg-preferential — all {n_sig_up} up arms clear "
                  f"FDR {fdr}")

        ax.axvline(0, ls="--", c="grey", lw=1)
        ax.set_yticks(yv); ax.set_yticklabels(ylabels)
        ax.set_xlim(0, 4.5)
        ax.set_xticks([0, 1, 2, 3])
        ax.set_xlabel("WT_heat NES (SF synovial fluid vs PB paired peripheral blood)")
        ax.set_title("QUESTION: does the mouse 39 °C-derived signature separate the inflamed\n"
                     "synovial niche from paired blood within a frozen sort label?\n"
                     f"ANSWER: yes — {answer}")
        from matplotlib.lines import Line2D
        # Legend has two keys: colour = population (matches the point colours), shape = up/down set.
        pop_handles = [
            Line2D([0], [0], marker="o", color="w", markerfacecolor=POP_COL[p],
                   markeredgecolor="k", label=p, markersize=9)
            for p in POP_TAG
        ]
        shape_handles = [
            Line2D([0], [0], marker="o", color="w", markerfacecolor="white",
                   markeredgecolor="k", label="up set (circle)", markersize=9),
            Line2D([0], [0], marker="D", color="w", markerfacecolor="white",
                   markeredgecolor="k", label="down set (diamond)", markersize=9),
            Line2D([0], [0], marker="", linestyle="", label=f"* = FDR < {fdr}"),
        ]
        # Upper right: the three down-arm rows sit at low NES, so that corner is the
        # only region no marker or size annotation occupies.
        ax.legend(handles=pop_handles + shape_handles, frameon=True, loc="upper right",
                  fontsize=9, title="population (colour) · set (shape)",
                  title_fontsize=9)
        # Two readings the sequence used to leave to prose. The Treg-preference
        # answer is negative; and the down arm is not silent, so "the up arm is
        # the only informative arm" is retired here rather than in a caption.
        down_note = (
            "The down arm is not silent either: it reaches FDR "
            + ", ".join(f"{float(r['padj']):.3f} in {r['cell_state']}"
                        for _, r in down_sig.iterrows())
            + " at the same sign as the up arm, and carries no direction in the other two.\n"
        ) if len(down_sig) else "No down arm reaches significance in any population.\n"
        ax.text(0.0, -0.135,
                "ANSWER — confirmatory tier (donor-level pseudobulk within frozen sort labels, "
                "limma-voom → fgsea). This is the only tier that may answer.\n"
                "Treg preference is answered in the NEGATIVE: Tregs are in the result, not "
                "privileged in it, and the effective set size tracks the NES\n"
                "across these three rows, so do not read the ordering as a biological ranking. "
                "'Treg signature' names how the set was derived — from mouse\n"
                "iTreg 37/39 °C contrasts — and never a specificity it does not have.\n"
                + down_note
                + "Geometry: ordered NES dot plot with FDR encoding; no pseudobulk NES row "
                  "carries an interval.",
                transform=ax.transAxes, ha="left", va="top", fontsize=9)
    fig.tight_layout()

    # Every NES, FDR and set size the caption states is read from `gsea` in this run. The
    # mouse arms are re-derived upstream in mouse_anchor, so a typed triple here would go
    # stale the moment that correction lands.
    def _arm_rows(direction: str) -> pd.DataFrame:
        if gsea.empty:
            return gsea
        sub = gsea[gsea["direction"].eq(direction)].copy()
        sub["ord"] = sub["cell_state"].map({p: i for i, p in enumerate(POP_TAG)})
        return sub.sort_values("ord")

    def _phrase(sub: pd.DataFrame, first_full: bool = True) -> str:
        bits = []
        for k, (_, r) in enumerate(sub.iterrows()):
            size = (f" ({int(r['set_size'])} of {int(r['n_nominal'])} genes ranked)"
                    if k == 0 and first_full else f" ({int(r['set_size'])})")
            bits.append(f"{float(r['nes']):.4f} in {r['cell_state']}{size}")
        return ", ".join(bits[:-1]) + (" and " if len(bits) > 1 else "") + bits[-1]

    up_rows, down_rows = _arm_rows("up"), _arm_rows("down")
    if gsea.empty:
        up_line = down_line = "No enrichment table was found for this stage."
    else:
        up_line = (f"The mouse 39 °C-derived up arm separates synovial fluid from paired blood "
                   f"in every sorted population: NES {_phrase(up_rows)}, all at FDR below "
                   f"{up_rows['padj'].max():.0e}. The result is pan-T, and Tregs are one of the "
                   f"three populations carrying it.")
        d_sig = down_rows[down_rows["padj"].lt(fdr)]
        d_ns = down_rows[~down_rows["padj"].lt(fdr)]
        down_line = (
            (f"The down arm reaches NES "
             + ", ".join(f"{float(r['nes']):.4f} at FDR {float(r['padj']):.3f} in {r['cell_state']}"
                         for _, r in d_sig.iterrows())
             + ", the same sign as the up arm, and carries no direction in "
             + " or ".join(f"{r['cell_state']} ({float(r['nes']):.4f})"
                           for _, r in d_ns.iterrows()) + ".")
            if len(d_sig) else
            f"The down arm reaches significance in no population at FDR {fdr}.")

    save_overview(fig, STAGE, "wt_heat_nes_forest",
                  table=gsea[["cell_state", "pathway_id", "nes", "pvalue", "padj",
                              "set_size", "n_nominal"]]
                  if not gsea.empty else pd.DataFrame(),
                  finding=f"{up_line} {down_line}",
                  script=SCRIPT, fn="main",
                  config_kv=f"thresholds.gsea_fdr={PARAMS.gsea_fdr}; gsea_min_size={PARAMS.gsea_min_size}",
                  input="03_results/05_scoring/tables/gsea_pseudobulk_{treg,tcon,cd8}.csv",
                  how_to_read=("This is the confirmatory tier: donor-level pseudobulk within "
                               "frozen sort labels, limma-voom then fgsea, on the 6 donors "
                               "present in both arms. Points are NES for the up (circle) and down "
                               "(diamond) arms, coloured by population; the asterisk marks FDR "
                               f"below {PARAMS.gsea_fdr}. Labels give effective and nominal set "
                               "sizes plus FDR. Effective size tracks the NES ordering, so the "
                               "ordering is a size effect. Ordered NES dot plot with FDR "
                               "encoding; the rows carry no interval. Correlative."),
                  config=FIG_CFG, width=9.5, height=8.5)

    # ---- 2. per-cell score violins ----
    dm = pd.read_csv(tdir / "donor_label_score_means.csv")
    dm["tissue_s"] = dm["tissue"].map({"synovial_fluid": "SF", "peripheral_blood": "PB"})
    dm["group"] = dm["coarse_label"] + " " + dm["tissue_s"]
    groups = ["Treg SF", "Treg PB", "Tcon SF", "Tcon PB", "CD8 SF", "CD8 PB"]
    fig3, ax = plt.subplots(figsize=(9, 6))
    data, positions, colors = [], [], []
    for i, grp in enumerate(groups):
        vals = dm.loc[dm["group"] == grp, "WT_heat_up_AUCell"].values
        if len(vals):
            data.append(vals); positions.append(i)
            colors.append(POP_COL[grp.split()[0]])
    parts = ax.violinplot(data, positions=positions, showmeans=True, widths=0.8)
    for b, c in zip(parts["bodies"], colors):
        b.set_facecolor(c); b.set_alpha(0.65)
    for i, grp in zip(positions, [groups[p] for p in positions]):
        vals = dm.loc[dm["group"] == grp, "WT_heat_up_AUCell"].values
        ax.scatter(np.full(len(vals), i), vals, s=18, c="black", zorder=3)
    ax.set_xticks(range(len(groups))); ax.set_xticklabels(groups, rotation=30, ha="right")
    ax.set_ylabel("donor-mean WT_heat_up AUCell score")
    ax.set_title("Per-cell WT_heat_up AUCell score, donor means by state × tissue\n"
                 "CORROBORATES ONLY — a per-cell score may never answer")
    fig3.tight_layout()
    save_overview(fig3, STAGE, "score_violins", table=dm,
                  finding=("Donor-mean WT_heat_up AUCell sits higher in synovial fluid than in "
                           "paired blood in all three sorted populations, so the per-cell channel "
                           "shadows the pseudobulk result in the same direction. This corroborates. "
                           "A per-cell score is a different estimand on a secondary tier, and the "
                           "shift it shows spans all three populations."),
                  script=SCRIPT, fn="main",
                  config_kv="signature = WT_heat_up (AUCell, rank-based [0,1])",
                  input="03_results/05_scoring/tables/donor_label_score_means.csv",
                  how_to_read=("This panel corroborates; the confirmatory answer is the pseudobulk "
                               "NES dot plot. Each dot is one donor's mean WT_heat_up AUCell score "
                               "for that state × tissue, and the violins summarise across donors. "
                               "AUCell is a rank-based score in [0, 1], the area under each cell's "
                               "gene-recovery curve for the up set, robust to library size and "
                               "composition. Read the SF-versus-PB shift within a population (Treg "
                               "SF against Treg PB); the absolute level carries no reading. This "
                               "is a different estimand from the pseudobulk NES dot plot and "
                               "shares no axis with it. The down arm is omitted because up and "
                               "down co-shift in synovial fluid. Correlative."),
                  config=FIG_CFG)
    print("[05_scoring_viz] wrote 2 overviews (ordered NES dot plot + score violins)")


if __name__ == "__main__":
    main()

File: 02_analysis/scripts/05_score_signatures_viz.R
#!/usr/bin/env Rscript
# 05_score_signatures_viz.R — VIZ (no statistics). Per-population running-sum.
# =====================================================================
# Drives the CANONICAL RNAseq-toolkit running-sum plotter
# (01_modules/RNAseq-toolkit/scripts/GSEA/GSEA_plotting/gsea_running_sum_plot.R)
# off the clusterProfiler `gseaResult` S4 objects written by the compute step
# (helpers/fgsea_prerank.R -> gsea_pseudobulk_{tag}.rds). The plotter overlays
# N gene SETS from ONE ranked list, so we emit ONE figure PER POPULATION, each
# overlaying WT_heat_up AND WT_heat_down. (The retired hand-rolled 3-population
# single-axis ggplot is gone: it cannot mix a set-overlay with a pop-overlay.)
#
# Input  (03_results/05_scoring/tables/):
#   gsea_pseudobulk_{treg,tcon,cd8}.rds   ← clusterProfiler gseaResult S4
#
# Output (03_results/05_scoring/):
#   figures/_overview/wt_heat_running_sum_{treg,tcon,cd8}.{pdf,png}
#   tables/_overview/wt_heat_running_sum_{treg,tcon,cd8}.csv   ← same-stem source table
#   README.md captions (via save_overview)
#
# Run from the compartment root, AFTER 05_score_signatures.py:
#   Rscript 02_analysis/scripts/05_score_signatures_viz.R

suppressPackageStartupMessages({
  library(ggplot2)
  library(patchwork)
})

source("02_analysis/helpers/figure_style.R")
source("02_analysis/helpers/source_hash_manifest.R")
source("01_modules/RNAseq-toolkit/scripts/GSEA/GSEA_plotting/gsea_running_sum_plot.R")

STAGE  <- "05_scoring"
SCRIPT <- "02_analysis/scripts/05_score_signatures_viz.R"
TDIR   <- "03_results/05_scoring/tables"

POP_TAG    <- c(Treg = "treg", Tcon = "tcon", CD8 = "cd8")
GENE_SETS  <- c("WT_heat_up", "WT_heat_down")
SET_LABELS <- c(WT_heat_up = "WT_heat up", WT_heat_down = "WT_heat down")
SET_ARM    <- c(WT_heat_up = "up", WT_heat_down = "down")
# Diverging cue: heat-up = warm = brown, heat-down = cool = blue (semantic,
# theme-consistent). Keyed BY LEGEND LABEL, which is the value the colour
# aesthetic actually carries — see `apply_set_palette()` for why a positional
# vector cannot express this mapping safely.
SET_PAL <- c("WT_heat up" = "#A6611A", "WT_heat down" = "#2166AC")

#' Re-assert the up/down colour mapping BY NAME on every panel.
#'
#' Defence in depth, kept deliberately. The original defect was upstream: the
#' toolkit plotter forwarded `palette` UNNAMED into `enrichplot::gseaplot2()`,
#' which maps `colour = Description` and calls `scale_color_manual(values =)`.
#' Unnamed values match by POSITION against levels ggplot has sorted
#' ALPHABETICALLY — "WT_heat down" precedes "WT_heat up" — so a positional
#' palette in GENE_SETS order put brown on the down curve. The `labels` lookup
#' is named, so the legend TEXT stayed correct while the colours swapped, and it
#' read as a deliberate bad choice rather than a bug.
#'
#' The toolkit now re-keys the palette to the plotted label itself and owns the
#' scale (RNAseq-toolkit `fix(GSEA): key running-sum palette by plotted label`),
#' so this is no longer load-bearing — verified idempotent: toolkit-only and
#' toolkit-plus-this both bind brown to the up curve. It stays because the
#' parent pins the toolkit by gitlink and that fix carries no release tag yet,
#' so an older pin would silently reintroduce the swap. Retire it once the
#' compartment pins a toolkit tag containing the fix. `limits` fixes legend
#' order to up-then-down.
apply_set_palette <- function(p, ids, labs) {
  labs_in_order <- unname(labs[ids])
  pal <- stats::setNames(unname(SET_PAL[SET_LABELS[ids]]), labs_in_order)
  suppressMessages(
    p & ggplot2::scale_color_manual(values = pal, limits = labs_in_order,
                                    name = NULL))
}

#' Nominal size of each frozen mouse arm — a line count, not a statistic.
#'
#' An effective set size only means something against the nominal one, so the
#' legend carries both. Read from the frozen mouse->human projection contract that
#' the compute step scored, so the denominator cannot drift from the numerator.
nominal_set_sizes <- function() {
  contract <- FIG_CFG$paths$signature_contract %||%
    "../mouse_anchor/03_results/human_projection/"
  sig_dir <- file.path(contract, "signatures", "WT_heat")
  vapply(GENE_SETS, function(id) {
    path <- file.path(sig_dir, paste0(id, ".txt"))
    if (!file.exists(path))
      stop("[05_viz_R] frozen signature not found: ", path)
    verify_source_hash(
      path, id, file.path(TDIR, "source_hash_manifest.csv"),
      root = normalizePath("..", mustWork = FALSE))
    genes <- trimws(readLines(path, warn = FALSE))
    length(unique(genes[nzchar(genes)]))
  }, FUN.VALUE = integer(1))
}

#' Render an FDR for an in-figure label: fixed below 3 decimals, else scientific.
fmt_fdr <- function(p) if (is.na(p)) "FDR n/a" else if (p >= 0.001)
  sprintf("FDR %.3f", p) else sprintf("FDR %.0e", p)

ylim    <- as.numeric(unlist(FIG_CFG$figures$running_sum_ylim     %||% c(-1, 1)))
heights <- as.numeric(unlist(FIG_CFG$figures$running_sum_heights  %||% c(2.4, 0.7, 0.9)))
base_thm <- if (exists("project_theme")) project_theme(config = FIG_CFG) else theme_classic(base_size = 14)

# Own the running-sum figure namespace: drop the retired single 3-population
# figure (wt_heat_running_sum.*) and any stale per-pop files before rewriting.
if (exists("purge_figures"))
  purge_figures(STAGE, "wt_heat_running_sum", overview = TRUE, config = FIG_CFG)

#' Build this population's finding from THIS population's own summary rows.
#'
#' A caption may quote only its same-stem table, and this figure's same-stem table
#' holds one population. The retired caption quoted all three populations' NES from
#' the dot plot's table instead, which made a three-population claim from a
#' one-population panel; the cross-population comparison belongs on the dot plot
#' and this caption points there rather than restating it.
finding_for <- function(pop, tbl, nominal) {
  parts <- vapply(seq_len(nrow(tbl)), function(i) {
    r <- tbl[i, ]
    sprintf("the %s arm reaches NES %+.4f at %s with %d of its %d genes in the ranked list",
            SET_ARM[[r$ID]], r$NES, fmt_fdr(r$p.adjust), r$setSize, nominal[[r$ID]])
  }, FUN.VALUE = character(1))
  paste0(
    "In ", pop, " ", paste(parts, collapse = ", and "), ". ",
    "The curve gives the place along this population's synovial-fluid-versus-blood ",
    "ranking where each arm concentrates. The cross-population comparison — whether ",
    "one sorted population separates more than another — is read off the ordered NES ",
    "dot plot.")
}

HOW_TO_READ <- paste(
  "One population per panel, showing the donor-pseudobulk fgsea result behind the",
  "confirmatory answer. The top trace walks from SF-enriched to PB-enriched genes;",
  "a positive left peak indicates SF enrichment. The middle rug marks set members",
  "and the bottom shows the signed moderated-t ranking. Warm brown is the up arm and",
  "cool blue the down arm. Legends report effective and nominal size, NES, and FDR.",
  "The shared [-1, 1] enrichment-score range supports shape comparison. Read the",
  "cross-population result from the ordered NES dot plot, which establishes the",
  "pan-T pattern. Display of compute output; correlative.")
CONFIG_KV <- paste0("gsea_min_size=", FIG_CFG$thresholds$gsea_min_size %||% 5,
                    "; gsea_max_size=", FIG_CFG$thresholds$gsea_max_size %||% 500,
                    "; running_sum_ylim=[", ylim[1], ",", ylim[2], "]",
                    "; engine=clusterProfiler::GSEA(by=fgsea)")

NOMINAL <- nominal_set_sizes()

n_written <- 0L
for (pop in names(POP_TAG)) {
  tag <- POP_TAG[[pop]]
  rds <- file.path(TDIR, sprintf("gsea_pseudobulk_%s.rds", tag))
  if (!file.exists(rds)) {
    warning(sprintf("[05_viz_R] gseaResult RDS not found, skipping %s: %s", pop, rds))
    next
  }
  g   <- readRDS(rds)
  ids <- intersect(GENE_SETS, g@result$ID)
  if (length(ids) == 0) {
    warning(sprintf("[05_viz_R] no WT_heat sets in %s result, skipping", pop))
    next
  }

  # Source table = the gseaResult summary rows behind THIS figure.
  res <- g@result
  tbl <- res[res$ID %in% ids,
             c("ID", "Description", "setSize", "enrichmentScore", "NES",
               "pvalue", "p.adjust", "core_enrichment")]
  tbl$population  <- pop
  tbl$n_nominal   <- unname(NOMINAL[tbl$ID])
  tbl <- tbl[match(ids, tbl$ID), ]

  # Effective set size travels with every NES, on the FACE. The legend label is the
  # only string on this plot wide enough to hold it, and the toolkit wraps labels
  # only past `max_name_length`, so the limit is raised past the longest label
  # rather than left to truncate one.
  labs <- stats::setNames(
    vapply(ids, function(id) {
      r <- tbl[tbl$ID == id, ][1, ]
      sprintf("%s   %d of %d genes ranked,  NES %+.2f,  %s",
              SET_LABELS[[id]], r$setSize, r$n_nominal, r$NES, fmt_fdr(r$p.adjust))
    }, FUN.VALUE = character(1)),
    ids)

  p <- gsea_running_sum_plot(
    g,
    gene_set_ids    = ids,
    palette         = stats::setNames(unname(SET_PAL[SET_LABELS[ids]]), ids),
    labels          = labs,
    max_name_length = max(nchar(labs)) + 1L,
    es_ylim         = ylim,
    panel_heights   = heights,
    legend_position = "right",
    base_theme      = base_thm,
    title           = sprintf(
      "%s — where the mouse 39 °C-derived arms sit\nalong this population's SF-vs-PB ranking",
      pop))
  p <- apply_set_palette(p, ids, labs)

  save_overview(
    p, STAGE, sprintf("wt_heat_running_sum_%s", tag), table = tbl,
    finding = finding_for(pop, tbl, NOMINAL), script = SCRIPT, fn = "main",
    config_kv = CONFIG_KV,
    input = sprintf("03_results/05_scoring/tables/gsea_pseudobulk_%s.rds", tag),
    how_to_read = HOW_TO_READ,
    config = FIG_CFG, width = 11, height = 7)
  n_written <- n_written + 1L
}

if (n_written == 0L)
  stop("[05_viz_R] No gseaResult RDS found in ", TDIR, ". Run 05_score_signatures.py first.")
message(sprintf("[05_viz_R] wrote %d per-population running-sum overview(s)", n_written))
